website/i18n/zh/docusaurus-plugin-content-blog/2022/01/21/apisix-hashicorp-vault-integration.md

```diff
@@ -0,0 +1,383 @@
+
+```shell
+$ curl http://127.0.0.1:9080/apisix/plugin/jwt/sign\?key\=test-key -i
+HTTP/1.1 200 OK
+Date: Tue, 18 Jan 2022 07:50:57 GMT
+Content-Type: text/plain; charset=utf-8
+Transfer-Encoding: chunked
+Connection: keep-alive
+Server: APISIX/2.11.0
+
+
+eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJrZXkiOiJ0ZXN0LWtleSIsImV4cCI6MTY0MjU3ODY1N30.nkyev1_KUapVgY_QVYETsSApA6gEkDWS8tsHFV1EpD8
+```
+
+在上一步中，如果你看到类似 `failed to sign jwt` 的信息，请确保你有一个私有密钥存储在 vault `kv/apisix/consumers/jack/jwt-aut` 路径中。
+
+```shell
+# example
+$ vault kv put kv/apisix/consumer/jack/jwt-auth secret=$ecr3t-c0d3
+Success! Data written to: kv/apisix/consumer/jack/jwt-auth
+```
+
+#### 向 APISIX Server 发送请求
+
+现在，向 APISIX 代理发出一个路由 `/secure/ping` 的请求。验证成功后，它将把请求转发给我们的 go HTTP server。
+
+```shell
+$ curl http://127.0.0.1:9080/secure/ping -H 'Authorization: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJrZXkiOiJ0ZXN0LWtleSIsImV4cCI6MTY0MjU3ODU5M30.IYudBr7FTgRme70u4rEBoYNtGmGByzgfGlt8hctI__Q' -i
+HTTP/1.1 200 OK
+Content-Type: text/plain; charset=utf-8
+Content-Length: 12
+Connection: keep-alive
+Date: Tue, 18 Jan 2022 08:00:04 GMT
+Server: APISIX/2.11.0
+
+
+secure/pong
+```
+
+任何无效的 JWT 请求都会抛出 `HTTP 401 Unauthorized` 的错误。
+
+```shell
+$ curl http://127.0.0.1:9080/secure/ping -i
+HTTP/1.1 401 Unauthorized
+Date: Tue, 18 Jan 2022 08:00:33 GMT
+Content-Type: text/plain; charset=utf-8
+Transfer-Encoding: chunked
+Connection: keep-alive
+Server: APISIX/2.11.0
+
+
+{"message":"Missing JWT token in request"}
+```
+
+### Vault 与 APISIX jwt-auth 插件集成的不同用例
+
+Apache APISIX `jwt-auth` 插件可以被配置为从 Vault 存储中获取简单的文本密钥以及 RS256 公私密钥对。
+
+:::note
+对于该集成支持的早期版本，该插件希望存储到金库路径中的密钥名称在 [`secret`, `public_key`, `private_key`] 之间，以成功使用该密钥。在未来的版本中，我们将增加对引用自定义命名的密钥的支持。
+:::
+
+1. 你在 Vault 内存储了 HS256 签名密钥，你想用它来进行 jwt 签名和验证。
+
+   ```shell
+   $ curl http://127.0.0.1:9080/apisix/admin/consumers -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+    {
+        "username": "jack",
+        "plugins": {
+            "jwt-auth": {
+                "key": "key-1",
+                "vault": {}
+            }
+        }
+    }'
+   ```
+
+   在这里，插件在 Consumer 配置中提到的 Consumer 用户 `jack` 的 Vault 路径（`<vault.prefix from conf.yaml>/consumer/jack/jwt-auth`）中查找密钥 `secret`，并使用它进行后续的签名和 jwt 验证。如果在同一路径中没有找到密钥，该插件将记录一个错误，并且无法执行 JWT 验证。
+
+2. RS256 RSA 密钥对，公钥和私钥都存储在 Vault 中。
+
+   ```shell
+   $ curl http://127.0.0.1:9080/apisix/admin/consumers -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+    {
+        "username": "jim",
+        "plugins": {
+            "jwt-auth": {
+                "key": "rsa-keypair",
+                "algorithm": "RS256",
+                "vault": {}
+            }
+        }
+    }'
+   ```
+
+   The plugin looks up for `public_key` and `private_key` keys inside vault kv path (`<vault.prefix from conf.yaml>/consumer/jim/jwt-auth`) for `jim` mentioned inside plugin vault configuration. If not found, authentication fails.
```

Comment:
```suggestion

```
Excess?

Reply:
solved in https://github.com/apache/apisix-website/pull/857